### ISSUE TITLE
[api-extractor] Make processProject() API more useful for wrapping api-extractor in other scripts

### ISSUE THREAD

**Issue by skaapgif** [Microsoft/web-build-tools#1133]
The boolean return type of processProject https://github.com/Microsoft/web-build-tools/blob/master/apps/api-extractor/src/api/Extractor.ts#L364 makes it hard for a wrapping script to get visibility into the results from api-extractor.

It would be more useful if processProject returned all errors and warnings that were encountered. That way the wrapping script could decide how to proceed.

A workaround is to create a customLogger that saves all log messages and ignoring processProject's return value. Since this is very similar to MonitoredLogger it feels like this should rather be exposed properly.

For more context, this is how we're currently working around this issue inside Kibana: https://github.com/elastic/kibana/blob/d45d2898d8ffb941573458dd826420d8b8cf5b6f/src/dev/run_check_core_api_changes.ts#L102

**Comment by octogonz** [comment]
This is a good suggestion and an easy fix.  I'll see if we can get it in before AE7 is released.

It would be more useful to return the array of `ExtractorMessage` objects, which provide more rich information that you can use for sorting/filtering.  The `ILogger` interface only provides text strings, since it's also used for general informational messages.

**Comment by skaapgif** [comment]
Thanks, ExtractorMessage objects would be even more useful and less brittle
than comparing string output.

**Comment by octogonz** [comment]
@skaapgif btw if you'd like an API Extractor t-shirt, ping me on Gitter.im :-)

**Comment by octogonz** [comment]
@skaapgif thanks very much for sharing the link to your project repo. I hadn't seen an example before of someone wrapping up the extractor/documenter functionality behind a different command-line interface.  That's pretty cool!

I'm working on a PR that will eliminate the `ILogger` interface entirely, and instead provide a list of `ExtractorMessage` objects for the caller to process.  (You might also be interested in [PR 1206](https://github.com/Microsoft/web-build-tools/pull/1206) which redesigns the `Extractor` API that you guys are using.)

**Comment by octogonz** [comment]
**Proposed design:**

I'm thinking that every possible thing that API Extractor might print to the console should be modeled as an [ExtractorMessage](https://github.com/Microsoft/web-build-tools/blob/master/apps/api-extractor/src/api/ExtractorMessage.ts).  This means each message will have a unique ID that you can use to identify it.  Also note that [IExtractorMessageProperties](https://github.com/Microsoft/web-build-tools/blob/9f41072d99d4d333e697c5bc992597a59f42ce5c/apps/api-extractor/src/api/ExtractorMessage.ts#L116) allows us to attach extra details to messages, e.g. for someone who wants to generate their own custom message that contains the same information.

And I'm thinking we'll aim to support 4 different ways of handling messages:
1. **Simple default behavior:** By default the messages go straight to stderr/stdout as before.  This keeps the API usage simple for people who are just getting started.
2. **Ability to customize text:** There will be an optional callback (`onMessage()`?) that allows you to customize the text before it goes to API Extractor's internal stderr/stdout logger.
3. **Ability to filter messages:** If you want to remove a message entirely, the callback can set `handled=true`.
4. **Ability to do custom handling:** You can also set `handled=true` for ALL messages, and process them in a completely custom way.  If you do this, API Extractor will never write anything to the console.

Some ideas that I considered but decided NOT to include:
- **A non-realtime mode:** (not doing this) `ExtractorResult` could return an array of messages after then (synchronous) invocation completes.  This avoids tools needing to implement a callback, but I think it's pretty easy for a caller to accumulate their own array using the callback.
- **Injection of custom messages:** (not doing this) The callback could provide a way to generate your own messages and inject them into the stream.  Not worth the complexity this would introduce.
- **Expose the steps of the algorithm:** (later) It would be nice to expose the various stages of API Extractor's processing into separate methods that allow the caller more control over what happens. We'll probably do this eventually, but I don't think it's needed right now.

**Comment by skaapgif** [comment]
For our specific use case, we wrap api-extractor just to find out if the api was changed (and if there were > 0 warnings or errors). After #1120 was merged, we now have a lot of control over messages. If `ExtractorResult` exposed `apiChanged: boolean` as a property we wouldn't have a need for an `onMessage` callback.

Apart from our use case, I do believe having control over the output with an onMessage handler would be useful. I agree that it's easy enough for a script to collect all messages so this doesn't need to be included in the `ExtractorResult`.

Just to be sure, you're suggesting an api like
```ts
onMessage = function(message: ExtractorMessage): ExtractorMessage
```

Where `ExtractorMessage` has the additional `handled: boolean` property?

**Comment by octogonz** [comment]
> If `ExtractorResult` exposed `apiChanged: boolean` as a property we wouldn't have a need for an `onMessage` callback.

Good suggestion -- we'll definitely add this.

**Comment by octogonz** [comment]
@skaapgif your example also makes me think we should provide a config setting that allows you to customize the text that is shown when the API signature changes.  For many people this might avoid the need to write script code at all.  I opened issue https://github.com/Microsoft/web-build-tools/issues/1207 to track this.

**Comment by octogonz** [comment]
@skaapgif  Right, I'm thinking we will add two properties to `ExtractorMessage`:

```ts
export class ExtractorMessage {
  . . .

  /**
   * If the `messageCallback` sets this property to true, it will prevent the message from being displayed
   * by API Extractor.
   *
   * @remarks
   * If the `messageCallback` routes the message to a custom handler (e.g. a toolchain logger), it should
   * assign `handled = true` to prevent API Extractor from displaying it.  Assigning `handled = true` for all messages
   * would effectively disable all console output from the `Extractor` API.
   *
   * If `handled` is set to true, the message will still be included in the count of errors/warnings;
   * to discard a message entirely, instead assign `logLevel = none`.
   */
  public handled: boolean;

  /**
   * Specifies how the message should be reported.
   *
   * @remarks
   * If the `messageCallback` handles the message (i.e. sets `handled = true`), it can use the `logLevel` to determine
   * how to display the message.
   *
   * Alternatively, if API Extractor is handling the message, the `messageCallback` can assign `logLevel` to change
   * how it will be processed.  But note that in general, the recommended practice is to configure message routing
   * using the `messages` section in api-extractor.json.
   *
   * To discard a message entirely, assign `logLevel = none`.
   */
  public logLevel: ExtractorMessageLogLevel;

  . . .
}
```

This way the callback is very simple:
```ts
export interface IExtractorInvokeOptions {
  . . .

  /**
   * An optional callback function that will be called for each `ExtractorMessage` before it is displayed by
   * API Extractor.  The callback can customize the message, handle it, or discard it.
   *
   * @remarks
   * If a `messageCallback` is not provided, then by default API Extractor will print the messages to
   * the STDERR/STDOUT console.
   */
  messageCallback: (message: ExtractorMessage) => void;
}
```

And the `ExtractorMessageLogLevel` will be expanded into include `Info` and `Verbose` values:
```ts
export const enum ExtractorMessageLogLevel {
  /**
   * The message will be displayed as an error.
   *
   * @remarks
   * Errors typically cause the build to fail and return a nonzero exit code.
   */
  Error = 'error',

  /**
   * The message will be displayed as an warning.
   *
   * @remarks
   * Warnings typically cause a production build fail and return a nonzero exit code.  For a non-production build
   * (e.g. using the `--local` option with `api-extractor run`), the warning is displayed but the build will not fail.
   */
  Warning = 'warning',

  /**
   * The message will be displayed as an informational message.
   *
   * @remarks
   * Informational messages may contain newlines to ensure nice formatting of the output,
   * however word-wrapping is the responsibility of the message handler.
   */
  Info = 'info',

  /**
   * The message will be displayed only when "verbose" output is requested, e.g. using the `--verbose`
   * command line option.
   */
  Verbose = 'verbose',

  /**
   * The message will be discarded entirely.
   */
  None = 'none'
}
```

I also suspect that [IConfigMessageReportingRule.addToApiReviewFile](https://github.com/Microsoft/web-build-tools/blob/ac161c4c4f9715e7cf62b23f128a971dcd5282ca/apps/api-extractor/src/api/IConfigFile.ts#L255) could be rolled into `ExtractorMessageLogLevel`, but I need to think about that that a little more.

**Comment by skaapgif** [comment]
> @skaapgif your example also makes me think we should provide a config setting that allows you to customize the text that is shown when the API signature changes. For many people this might avoid the need to write script code at all.

Yes, that would be great.

```ts
export interface IExtractorInvokeOptions {
  . . .

  /**
   * An optional callback function that will be called for each `ExtractorMessage` before it is displayed by
   * API Extractor.  The callback can customize the message, handle it, or discard it.
   *
   * @remarks
   * If a `messageCallback` is not provided, then by default API Extractor will print the messages to
   * the STDERR/STDOUT console.
   */
  messageCallback: (message: ExtractorMessage) => void;
}
```

Personally I'd prefer if `messageCallback` returned the new ExtractorMessage instead of directly mutating the parameter. Most people might end up mutating the parameter and returning that, making these two options technically nearly identical, but it feels to me like having a return value makes it more explicit that you're changing data that will alter the output of the application.

**Comment by octogonz** [comment]
@skaapgif here's the PR: https://github.com/Microsoft/web-build-tools/pull/1211

**Comment by octogonz** [comment]
This fix is published as API Extractor beta 7.0.38
